File: addons/terrabrush/src/Scripts/ImportExport/ImporterEngine.cs
#if TOOLS
using System;
using System.Collections.Generic;
using System.Linq;
using Godot;

namespace TerraBrush;

public partial class ImporterSettings : GodotObject {
    public Texture2D Heightmap { get;set; }
    public bool UseGreenChannelForHoles { get;set;}
    public bool ScaleToResolution { get;set;}
    public float HeightmapScale { get;set; }
    public Texture2D[] Splatmaps { get;set;}
    public Texture2D[] Foliages { get;set;}
    public Texture2D[] Objects { get;set;}
    public Texture2D Water { get;set;}
    public Texture2D Snow { get;set;}
}

public static class ImporterEngine {
    public static void ImportTerrain(TerraBrushTool terrabrush, ImporterSettings settings) {
        if (!string.IsNullOrWhiteSpace(terrabrush.DataPath) && !DirAccess.DirExistsAbsolute(terrabrush.DataPath)) {
            DirAccess.MakeDirAbsolute(terrabrush.DataPath);
        }

        terrabrush.TerrainZones??= new ZonesResource();
        terrabrush.TerrainZones.Zones??= new ZoneResource[] {};

        // Heightmap
        if (settings.Heightmap != null) {
            var resultImages = GenerateImageTextureForZones(
                terrabrush,
                settings.Heightmap.GetImage(),
                (zoneX, zoneY) => {
                    return ZoneUtils.CreateHeightmapImage(terrabrush.ZonesSize, terrabrush.Resolution, new Vector2I(zoneX, zoneY), terrabrush.DataPath);
                },
                (x, y, pixel, image) => {
                    var resultHeight = pixel.R * settings.HeightmapScale;
                    image.SetPixel(x, y, new Color(resultHeight, settings.UseGreenChannelForHoles ? pixel.G : 0, 0, 1));
                },
                true,
                settings.ScaleToResolution
            );

            foreach (var resultImage in resultImages) {
                GetZoneForImageInfo(terrabrush, resultImage).HeightMapTexture = resultImage.ImageTexture;
            }
        }

        // Splatmaps
        if (settings.Splatmaps?.Length > 0) {
            for (var i = 0; i < settings.Splatmaps.Length; i++) {
                var splatmap = settings.Splatmaps[i];

                var resultImages = GenerateImageTextureForZones(
                    terrabrush,
                    splatmap.GetImage(),
                    (zoneX, zoneY) => {
                        return ZoneUtils.CreateSplatmapImage(terrabrush.ZonesSize, new Vector2I(zoneX, zoneY), i, terrabrush.DataPath);
                    },
                    (x, y, pixel, image) => {
                        image.SetPixel(x, y, pixel);
                    }
                );

                foreach (var resultImage in resultImages) {
                    var zone = GetZoneForImageInfo(terrabrush, resultImage);
                    zone.SplatmapsTexture ??= new ImageTexture [] {};
                    if (zone.SplatmapsTexture.Length < i + 1) {
                        zone.SplatmapsTexture = zone.SplatmapsTexture.Append(resultImage.ImageTexture).ToArray();
                    } else {
                        zone.SplatmapsTexture[i] = resultImage.ImageTexture;
                    }
                }
            }
        }

        // Foliages
        if (settings.Foliages?.Length > 0) {
            for (var i = 0; i < settings.Foliages.Length; i++) {
                var foliage = settings.Foliages[i];

                var resultImages = GenerateImageTextureForZones(
                    terrabrush,
                    foliage.GetImage(),
                    (zoneX, zoneY) => {
                        return ZoneUtils.CreateFoliageImage(terrabrush.ZonesSize, new Vector2I(zoneX, zoneY), i, terrabrush.DataPath);
                    },
                    (x, y, pixel, image) => {
                        image.SetPixel(x, y, pixel);
                    }
                );

                foreach (var resultImage in resultImages) {
                    var zone = GetZoneForImageInfo(terrabrush, resultImage);
                    zone.FoliagesTexture ??= new ImageTexture [] {};
                    if (zone.FoliagesTexture.Length < i + 1) {
                        zone.FoliagesTexture = zone.FoliagesTexture.Append(resultImage.ImageTexture).ToArray();
                    } else {
                        zone.FoliagesTexture[i] = resultImage.ImageTexture;
                    }
                }
            }
        }

        // Objects
        if (settings.Objects?.Length > 0) {
            for (var i = 0; i < settings.Objects.Length; i++) {
                var objectItem = settings.Objects[i];

                var resultImages = GenerateImageTextureForZones(
                    terrabrush,
                    objectItem.GetImage(),
                    (zoneX, zoneY) => {
                        return ZoneUtils.CreateObjectImage(terrabrush.ZonesSize, new Vector2I(zoneX, zoneY), i, terrabrush.DataPath);
                    },
                    (x, y, pixel, image) => {
                        image.SetPixel(x, y, pixel);
                    }
                );

                foreach (var resultImage in resultImages) {
                    var zone = GetZoneForImageInfo(terrabrush, resultImage);
                    zone.ObjectsTexture ??= new ImageTexture [] {};
                    if (zone.ObjectsTexture.Length < i + 1) {
                        zone.ObjectsTexture = zone.ObjectsTexture.Append(resultImage.ImageTexture).ToArray();
                    } else {
                        zone.ObjectsTexture[i] = resultImage.ImageTexture;
                    }
                }
            }
        }

        // Water
        if (settings.Water != null) {
            var resultImages = GenerateImageTextureForZones(
                terrabrush,
                settings.Water.GetImage(),
                (zoneX, zoneY) => {
                    return ZoneUtils.CreateWaterImage(terrabrush.ZonesSize, terrabrush.Resolution, new Vector2I(zoneX, zoneY), terrabrush.DataPath);
                },
                (x, y, pixel, image) => {
                    image.SetPixel(x, y, pixel);
                },
                true,
                settings.ScaleToResolution
            );

            foreach (var resultImage in resultImages) {
                GetZoneForImageInfo(terrabrush, resultImage).WaterTexture = resultImage.ImageTexture;
            }
        }

        // Snow
        if (settings.Snow != null) {
            var resultImages = GenerateImageTextureForZones(
                terrabrush,
                settings.Snow.GetImage(),
                (zoneX, zoneY) => {
                    return ZoneUtils.CreateSnowImage(terrabrush.ZonesSize, terrabrush.Resolution, new Vector2I(zoneX, zoneY), terrabrush.DataPath);
                },
                (x, y, pixel, image) => {
                    image.SetPixel(x, y, pixel);
                },
                true,
                settings.ScaleToResolution
            );

            foreach (var resultImage in resultImages) {
                GetZoneForImageInfo(terrabrush, resultImage).SnowTexture = resultImage.ImageTexture;
            }
        }
    }

    private static ZoneResource GetZoneForImageInfo(TerraBrushTool terraBrush, ImportImageInfo imageInfo) {
        var zone = terraBrush.TerrainZones.Zones.FirstOrDefault(x => x.ZonePosition.X == imageInfo.ZoneX && x.ZonePosition.Y == imageInfo.ZoneY);
        if (zone == null) {
            zone = new ZoneResource() {
                ZonePosition = new Vector2I(imageInfo.ZoneX, imageInfo.ZoneY)
            };

            terraBrush.TerrainZones.Zones = new List<ZoneResource>(terraBrush.TerrainZones.Zones) {
                zone
            }.ToArray();
        }

        return zone;
    }

    private static List<ImportImageInfo> GenerateImageTextureForZones(TerraBrushTool terrabrush, Image image, Func<int, int, ImageTexture> generateNewImageCallback, Action<int, int, Color, Image> applyPixelToNewImage, bool applyResolution = false, bool scaleToResolution = true) {
        if (terrabrush.Resolution != 1 && applyResolution && scaleToResolution) {
            var newImage = new Image();
            newImage.CopyFrom(image);
            newImage.Resize(image.GetWidth() / terrabrush.Resolution, image.GetHeight() / terrabrush.Resolution);
            image = newImage;
        }

        var resolutionZoneSize = ZoneUtils.GetImageSizeForResolution(terrabrush.ZonesSize, applyResolution ? terrabrush.Resolution : 1);

        var xNumberOfZones = Mathf.CeilToInt(image.GetWidth() / (float) resolutionZoneSize);
        var yNumberOfZones = Mathf.CeilToInt(image.GetHeight() / (float) resolutionZoneSize);
        var resultList = new List<ImportImageInfo>();

        for (var x = 0; x < xNumberOfZones; x++) {
            for (var y = 0; y < yNumberOfZones; y++) {
                resultList.Add(GenerateImageTextureForZone(image, x, y, generateNewImageCallback, applyPixelToNewImage));
            }
        }

        return resultList;
    }

    private static ImportImageInfo GenerateImageTextureForZone(Image image, int zoneX, int zoneY, Func<int, int, ImageTexture> generateNewImageCallback, Action<int, int, Color, Image> applyPixelToNewImage) {
        var newImageTexture = generateNewImageCallback(zoneX, zoneY);
        var newImage = newImageTexture.GetImage();
        var newImageSize = newImage.GetWidth();

        var startingX = newImageSize * zoneX;
        var startingY = newImageSize * zoneY;
        for (var x = 0; x < newImageSize; x++) {
            for (var y = 0; y < newImageSize; y++) {
                var toX = x + startingX;
                var toY = y + startingY;

                var pixel = Colors.Black;

                // Try to match the next pixel with the one of the previous zone, for better transition
                if (x == newImageSize - 1) {
                    toX += 1;
                }

                if (y == newImageSize - 1) {
                    toY += 1;
                }

                if (toX < image.GetWidth() && toY < image.GetHeight()) {
                    pixel = image.GetPixel(toX, toY);
                }

                applyPixelToNewImage(x, y, pixel, newImage);
            }
        }

        if (!string.IsNullOrWhiteSpace(newImageTexture.ResourcePath)) {
            ResourceSaver.Save(newImageTexture);
        }

        return new ImportImageInfo {
            ZoneX = zoneX,
            ZoneY = zoneY,
            ImageTexture = newImageTexture
        };
    }

    internal class ImportImageInfo {
        public int ZoneX { get;set; }
        public int ZoneY { get;set; }
        public ImageTexture ImageTexture { get;set; }
    }
}
#endif

File: src/misc/importer_engine.cpp
#include "importer_engine.h"
#include "../editor_resources/zones_resource.h"
#include "../editor_resources/zone_resource.h"
#include "../misc/zone_utils.h"

#include <godot_cpp/classes/dir_access.hpp>
#include <godot_cpp/classes/resource_saver.hpp>
#include <godot_cpp/classes/resource_loader.hpp>

using namespace godot;

void ImporterEngine::importTerrain(TerraBrush *terrabrush, ImporterSettings settings) {
    if (!terrabrush->get_dataPath().is_empty() && !DirAccess::dir_exists_absolute(terrabrush->get_dataPath())) {
        DirAccess::make_dir_absolute(terrabrush->get_dataPath());
    }

    if (terrabrush->get_terrainZones().is_null()) {
        terrabrush->set_terrainZones(memnew(ZonesResource));
    }

    // Heightmap
    if (!settings.heightmap.is_null()) {
        std::vector<ImportImageInfo> resultImages = generateImageTextureForZones(
            terrabrush,
            settings.heightmap->get_image(),
            ([terrabrush](int zoneX, int zoneY) {
                return ZoneUtils::createHeightmapImage(terrabrush->get_zonesSize(), terrabrush->get_resolution(), Vector2i(zoneX, zoneY), terrabrush->get_dataPath());
            }),
            ([settings](int x, int y, Color pixel, Ref<Image> image) {
                float resultHeight = pixel.r * settings.heightmapScale;
                image->set_pixel(x, y, Color(resultHeight, settings.useGreenChannelForHoles ? pixel.g : 0, 0, 1));
            }),
            true,
            settings.scaleToResolution
        );

        for (ImportImageInfo resultImage : resultImages) {
            getZoneForImageInfo(terrabrush, resultImage)->set_heightMapImage(resultImage.image);
        }
    }

    // Color
    if (!settings.color.is_null()) {
        std::vector<ImportImageInfo> resultImages = generateImageTextureForZones(
            terrabrush,
            settings.color->get_image(),
            ([terrabrush](int zoneX, int zoneY) {
                return ZoneUtils::createColorImage(terrabrush->get_zonesSize(), Vector2i(zoneX, zoneY), terrabrush->get_dataPath());
            }),
            ([settings](int x, int y, Color pixel, Ref<Image> image) {
                image->set_pixel(x, y, pixel);
            })
        );

        for (ImportImageInfo resultImage : resultImages) {
            getZoneForImageInfo(terrabrush, resultImage)->set_colorImage(resultImage.image);
        }
    }

    // Splatmaps
    if (settings.splatmaps.size() > 0) {
        for (int i = 0; i < settings.splatmaps.size(); i++) {
            Ref<Texture2D> splatmap = settings.splatmaps[i];

            std::vector<ImportImageInfo> resultImages = generateImageTextureForZones(
                terrabrush,
                splatmap->get_image(),
                ([terrabrush, i](int zoneX, int zoneY) {
                    return ZoneUtils::createSplatmapImage(terrabrush->get_zonesSize(), Vector2i(zoneX, zoneY), i, terrabrush->get_dataPath());
                }),
                ([](int x, int y, Color pixel, Ref<Image> image) {
                    image->set_pixel(x, y, pixel);
                })
            );

            for (ImportImageInfo resultImage : resultImages) {
                Ref<ZoneResource> zone = getZoneForImageInfo(terrabrush, resultImage);

                if (zone->get_splatmapsImage().size() < i + 1) {
                    zone->get_splatmapsImage().append(resultImage.image);
                } else {
                    zone->get_splatmapsImage()[i] = resultImage.image;
                }
            }
        }
    }

    // Foliages
    if (settings.foliages.size() > 0) {
        for (int i = 0; i < settings.foliages.size(); i++) {
            Ref<Texture2D> foliage = settings.foliages[i];

            std::vector<ImportImageInfo> resultImages = generateImageTextureForZones(
                terrabrush,
                foliage->get_image(),
                ([terrabrush, i](int zoneX, int zoneY) {
                    return ZoneUtils::createFoliageImage(terrabrush->get_zonesSize(), Vector2i(zoneX, zoneY), i, terrabrush->get_dataPath());
                }),
                ([](int x, int y, Color pixel, Ref<Image> image) {
                    image->set_pixel(x, y, pixel);
                })
            );

            for (ImportImageInfo resultImage : resultImages) {
                Ref<ZoneResource> zone = getZoneForImageInfo(terrabrush, resultImage);

                if (zone->get_foliagesImage().size() < i + 1) {
                    zone->get_foliagesImage().append(resultImage.image);
                } else {
                    zone->get_foliagesImage()[i] = resultImage.image;
                }
            }
        }
    }

    // Objects
    if (settings.objects.size() > 0) {
        for (int i = 0; i < settings.objects.size(); i++) {
            Ref<Texture2D> objectItem = settings.objects[i];

            std::vector<ImportImageInfo> resultImages = generateImageTextureForZones(
                terrabrush,
                objectItem->get_image(),
                ([terrabrush, i](int zoneX, int zoneY) {
                    return ZoneUtils::createObjectImage(terrabrush->get_zonesSize(), Vector2i(zoneX, zoneY), i, terrabrush->get_dataPath());
                }),
                ([](int x, int y, Color pixel, Ref<Image> image) {
                    image->set_pixel(x, y, pixel);
                })
            );

            for (ImportImageInfo resultImage : resultImages) {
                Ref<ZoneResource> zone = getZoneForImageInfo(terrabrush, resultImage);

                if (zone->get_objectsImage().size() < i + 1) {
                    zone->get_objectsImage().append(resultImage.image);
                }
                else {
                    zone->get_objectsImage()[i] = resultImage.image;
                }
            }
        }
    }

    // Water
    if (!settings.water.is_null()) {
        std::vector<ImportImageInfo> resultImages = generateImageTextureForZones(
            terrabrush,
            settings.water->get_image(),
            ([terrabrush](int zoneX, int zoneY) {
                return ZoneUtils::createWaterImage(terrabrush->get_zonesSize(), terrabrush->get_resolution(), Vector2i(zoneX, zoneY), terrabrush->get_dataPath());
            }),
            ([](int x, int y, Color pixel, Ref<Image> image) {
                image->set_pixel(x, y, pixel);
            }),
            true,
            settings.scaleToResolution
        );

        for (ImportImageInfo resultImage : resultImages) {
            getZoneForImageInfo(terrabrush, resultImage)->set_waterImage(resultImage.image);
        }
    }

    // Snow
    if (!settings.snow.is_null()) {
        std::vector<ImportImageInfo> resultImages = generateImageTextureForZones(
            terrabrush,
            settings.snow->get_image(),
            ([terrabrush](int zoneX, int zoneY) {
                return ZoneUtils::createSnowImage(terrabrush->get_zonesSize(), terrabrush->get_resolution(), Vector2i(zoneX, zoneY), terrabrush->get_dataPath());
            }),
            ([](int x, int y, Color pixel, Ref<Image> image) {
                image->set_pixel(x, y, pixel);
            }),
            true,
            settings.scaleToResolution
        );

        for (ImportImageInfo resultImage : resultImages) {
            getZoneForImageInfo(terrabrush, resultImage)->set_snowImage(resultImage.image);
        }
    }

    // MetaInfo
    if (!settings.metaInfo.is_null()) {
        std::vector<ImportImageInfo> resultImages = generateImageTextureForZones(
            terrabrush,
            settings.metaInfo->get_image(),
            ([terrabrush](int zoneX, int zoneY) {
                return ZoneUtils::createMetaInfoImage(terrabrush->get_zonesSize(), terrabrush->get_resolution(), Vector2i(zoneX, zoneY), terrabrush->get_dataPath());
            }),
            ([](int x, int y, Color pixel, Ref<Image> image) {
                image->set_pixel(x, y, pixel);
            }),
            true,
            settings.scaleToResolution
        );

        for (ImportImageInfo resultImage : resultImages) {
            getZoneForImageInfo(terrabrush, resultImage)->set_metaInfoImage(resultImage.image);
        }
    }
}

Ref<ZoneResource> ImporterEngine::getZoneForImageInfo(TerraBrush *terrabrush, ImportImageInfo imageInfo) {
    Ref<ZoneResource> zone = nullptr;

    for (Ref<ZoneResource> currentZone : terrabrush->get_terrainZones()->get_zones()) {
        if (currentZone->get_zonePosition().x == imageInfo.zoneX && currentZone->get_zonePosition().y == imageInfo.zoneY) {
            zone = currentZone;
            break;
        }
    }

    if (zone.is_null()) {
        zone = Ref<ZoneResource>(memnew(ZoneResource));
        zone->set_zonePosition(Vector2i(imageInfo.zoneX, imageInfo.zoneY));

        TypedArray<Ref<ZoneResource>> zones = terrabrush->get_terrainZones()->get_zones();
        zones.append(zone);
        terrabrush->get_terrainZones()->set_zones(zones);
    }

    return zone;
}

std::vector<ImportImageInfo> ImporterEngine::generateImageTextureForZones(TerraBrush *terrabrush, Ref<Image> image, std::function<Ref<Image>(int, int)> generateNewImageCallback, std::function<void(int, int, Color, Ref<Image>)> applyPixelToNewImage, bool applyResolution, bool scaleToResolution) {
    if (terrabrush->get_resolution() != 1 && applyResolution && scaleToResolution) {
        Ref<Image> newImage = memnew(Image);
        newImage->copy_from(image);
        newImage->resize(image->get_width() / terrabrush->get_resolution(), image->get_height() / terrabrush->get_resolution());
        image = newImage;
    }

    int resolutionZoneSize = ZoneUtils::getImageSizeForResolution(terrabrush->get_zonesSize(), applyResolution ? terrabrush->get_resolution() : 1);

    int xNumberOfZones = (int) Math::ceil(image->get_width() / (float) resolutionZoneSize);
    int yNumberOfZones = (int) Math::ceil(image->get_height() / (float) resolutionZoneSize);
    std::vector<ImportImageInfo> resultList = std::vector<ImportImageInfo>();

    for (int x = 0; x < xNumberOfZones; x++) {
        for (int y = 0; y < yNumberOfZones; y++) {
            resultList.push_back(generateImageTextureForZone(image, x, y, generateNewImageCallback, applyPixelToNewImage));
        }
    }

    return resultList;
}

ImportImageInfo ImporterEngine::generateImageTextureForZone(Ref<Image> image, int zoneX, int zoneY, std::function<Ref<Image>(int, int)> generateNewImageCallback, std::function<void(int, int, Color, Ref<Image>)> applyPixelToNewImage) {
    Ref<Image> newImage = generateNewImageCallback(zoneX, zoneY);
    int newImageSize = newImage->get_width();

    int startingX = newImageSize * zoneX;
    int startingY = newImageSize * zoneY;
    for (int x = 0; x < newImageSize; x++) {
        for (int y = 0; y < newImageSize; y++) {
            int toX = x + startingX;
            int toY = y + startingY;

            Color pixel = Color(0, 0, 0, 1);

            // Try to match the next pixel with the one of the previous zone, for better transition
            if (x == newImageSize - 1) {
                toX += 1;
            }

            if (y == newImageSize - 1) {
                toY += 1;
            }

            if (toX < image->get_width() && toY < image->get_height()) {
                pixel = image->get_pixel(toX, toY);
            }

            applyPixelToNewImage(x, y, pixel, newImage);
        }
    }

    if (!newImage->get_path().is_empty()) {
        ResourceSaver::get_singleton()->save(newImage, newImage->get_path());
        newImage = ResourceLoader::get_singleton()->load(newImage->get_path());
    }

    ImportImageInfo info = ImportImageInfo();
    info.zoneX = zoneX;
    info.zoneY = zoneY;
    info.image = newImage;
    return info;
}
